File: examples/quick_sort.rs
use std::{fmt::Debug, time::Instant};

fn main() {
    let mut arr = [34, 7, 23, 32, 5, 62, 1, 31, 32, 5];
    println!("Original array: {:?}", arr);
    let start = Instant::now();
    quicksort(&mut arr);
    let duration = start.elapsed();
    println!("Sorted array: {:?}", arr);
    println!("Time taken: {:?}", duration);
}

fn quicksort<T: Ord + Debug>(arr: &mut [T]) {
    if arr.len() > 1 {
        let pivot_index = partition(arr);
        // println!("array: {arr:?} pivot_index: {pivot_index}");
        let (left, right) = arr.split_at_mut(pivot_index);
        // println!("left: {left:?} right: {right:?}");
        quicksort(left);
        quicksort(&mut right[1..]);
    }
}

fn partition<T: Ord>(arr: &mut [T]) -> usize {
    let pivot_index = arr.len() / 2;
    arr.swap(pivot_index, arr.len() - 1);
    let mut store_index = 0;
    for i in 0..arr.len() - 1 {
        if arr[i] < arr[arr.len() - 1] {
            arr.swap(i, store_index);
            store_index += 1;
        }
    }
    arr.swap(store_index, arr.len() - 1);
    store_index
}

File: examples/quick_sort.cs
int[] arr = { 34, 7, 23, 32, 5, 62, 1, 31, 32, 5 };
Console.WriteLine("{0,-18} {1}", "Original array: ", String.Join(", ", arr));

var segment = new ArraySegment<int>(arr);
var start = Stopwatch.StartNew();
QuickSort(segment);
var duration = start.Elapsed;
Console.WriteLine("Time taken: " + duration);

Console.WriteLine("{0,-18} {1}", "Sorted array: ", String.Join(", ", arr));

void QuickSort<T>(ArraySegment<T> segment) where T : IComparable<T>
{
    if (segment.Count > 1)
    {
        var pivotIndex = Partition(segment);
        // 跳过已经排序好的索引
        var left = segment[0..pivotIndex];
        var right = segment[(pivotIndex + 1)..];

        // Console.WriteLine("Left array: " + String.Join(", ", left));
        // Console.WriteLine("Right array: " + String.Join(", ", right));

        QuickSort(left);
        QuickSort(right);
    }
}

int Partition<T>(ArraySegment<T> segment) where T : IComparable<T>
{
    var pivotIndex = segment.Count / 2;
    var lastIndex = segment.Count - 1;
    //将中间值移动到尾部
    Swap(segment, pivotIndex, lastIndex);

    var storeIndex = 0;
    for (var index = 0; index < lastIndex; index++)
    {
        if (segment[index].CompareTo(segment[lastIndex]) < 0)
        {
            Swap(segment, index, storeIndex);
            storeIndex++;
        }
    }

    //将中间值从尾部移动到排序后的位置
    Swap(segment, storeIndex, lastIndex);
    return storeIndex;
}

void Swap<T>(ArraySegment<T> segment, int i, int j)
{
    var tmp = segment[i];
    segment[i] = segment[j];
    segment[j] = tmp;
}